File: dropwizard-extra-common/src/main/scala/com/datasift/dropwizard/health/ZooKeeperHealthCheck.scala
package com.datasift.dropwizard.health

import com.yammer.metrics.core.HealthCheck
import com.yammer.metrics.core.HealthCheck.Result
import java.net.Socket

/** [[com.yammer.metrics.core.HealthCheck]] for a ZooKeeper Quorum */
class ZooKeeperHealthCheck(hosts: Set[String], port: Int)
  extends HealthCheck("ZooKeeper: " + hosts.map(_ + ":" + port).mkString(", ")) {

  override protected def check = {
    val connected = hosts filter {
      new Socket(_, port).isConnected
    }

    if (connected.isEmpty) {
      // total failure
      Result.unhealthy("Unable to connect to any nodes in quorum: " +
        hosts.mkString(", "))
    } else if (connected.size < hosts.size) {
      // partial failure
      Result.healthy("Unable to connect to some nodes in quorum: " +
        hosts.diff(connected).mkString(", "))
    } else {
      // everything's fine
      Result.healthy()
    }
  }
}

File: dropwizard-extra-core/src/main/java/com/datasift/dropwizard/health/ZooKeeperHealthCheck.java
package com.datasift.dropwizard.health;

import com.yammer.metrics.core.HealthCheck;

import java.net.Socket;

/**
 * A {@link HealthCheck} for a quorum of ZooKeeper nodes.
 */
public class ZooKeeperHealthCheck extends SocketHealthCheck {

    public ZooKeeperHealthCheck(final String host,
                                final int port,
                                final String name) {
        super(host, port, name);
    }

    /**
     * Checks that a ZooKeeper quorum node is operating correctly.
     *
     * @param socket a {@link Socket} connected to the ZooKeeper quorum node
     * @return Healthy if the ZooKeeper quorum node is operating correctly;
     *         otherwise, unhealthy
     */
    @Override
    protected Result check(final Socket socket) {
        // todo: find a way to verify the node is responding correctly
        return Result.healthy();
    }
}
